gadget/update.go

```diff
@@ -665,6 +667,143 @@ func DiskTraitsFromDeviceAndValidate(expLayout *LaidOutVolume, dev string, opts
+volumeLoop:
+	for volName, vol := range old.Info.Volumes {
+		for _, structure := range vol.Structure {
+			if structure.Role == SystemBoot {
+				// this is the volume
+				systemBootVolume = volName
+				break volumeLoop
+			}
+		}
+	}
+	if systemBootVolume == "" {
+		// didn't find system-boot anywhere somehow
+		if preUC20 {
+			logger.Noticef("WARNING: could not identify disk for gadget asset update of volume %s: unable to find any volume with system-boot role on it", systemBootVolume)
+			return nil, errSkipUpdateProceedRefresh
+		}
+		// shouldn't be possible on UC20
+		return nil, fmt.Errorf("unable to find any volume with system-boot, gadget is broken")
+	}
+
+	laidOutVol := laidOutVols[systemBootVolume]
+
+	// search for matching devices that correspond to the volume we laid out
+	dev := ""
+	for _, vs := range laidOutVol.LaidOutStructure {
+		// here it is okay that we require there to be either a partition label
+		// or a filesystem label since we require there to be a system-boot role
+		// on this volume which by definition must have a filesystem
+		structureDevice, err := FindDeviceForStructure(&vs)
+		if err == ErrDeviceNotFound {
+			continue
+		}
+		if err != nil {
+			// TODO: should this be a fatal error?
+			return nil, err
+		}
+
+		// we found a device for this structure, get the parent disk
+		// and save that as the device for this volume
+		disk, err := disks.DiskFromPartitionDeviceNode(structureDevice)
+		if err != nil {
+			// TODO: should we keep looping instead and try again with
+			// another structure? it probably wouldn't work because we found
+			// something on disk with the same name as something from the
+			// gadget.yaml, but then we failed to make a disk from that
+			// partition which suggests something is inconsistent with the
+			// state of the disk/udev database
+			return nil, err
+		}
+
+		dev = disk.KernelDeviceNode()
+		break
+	}
+
+	if dev == "" {
+		// couldn't find a disk at all, pre-UC20 we just warn about this
+		// but let the update continue
+		if preUC20 {
+			logger.Noticef("WARNING: could not identify disk for gadget asset update of volume %s", systemBootVolume)
+			return nil, errSkipUpdateProceedRefresh
+		}
+		// fatal error on UC20+
+		return nil, fmt.Errorf("could not identify disk for gadget asset update of volume %s", systemBootVolume)
+	}
+
+	// we found the device, construct the traits with validation options
+	validateOpts := &DiskVolumeValidationOptions{
+		// allow implicit system-data on pre-uc20 only
+		AllowImplicitSystemData: preUC20,
+	}
+
+	// setup encrypted structure information to perform validation if this
+	// device used encryption
+	if !preUC20 {
+		// TODO: this needs to check if the specified partitions are ICE when
+		// we support ICE too
+
+		// TODO: is this a good enough check to know if encryption was turned
+		// on? note that we can't import the boot pkg here
+		keysPattern := filepath.Join(dirs.SnapFDEDir, "*.key")
```

Comment:
I would use marker perhaps. Most of the .key will go away once we make recovery key optional, so at most we would have ubuntu-save.key

Reply:
changed to use the marker file instead